### ISSUE TITLE
Simplification of pruned models

### ISSUE THREAD

**Issue by carloalbertobarbano** [pytorch/pytorch#58846]
## 🚀 Feature

Extension of pytorch pruning utilities to effectively allow for the simplification of pruned models (i.e. actually removing zeroed out tensors from the model).

## Motivation

Pruning allows for impressive theoretical reduction of models sizes and complexity. However it usually offers little practical benefits as it is most often limited to just zeroing out weights, without actually removing the pruned filters/weights. This precludes from actual advantages provided by sparsification methods.

There seems to be a quite strong interest on the matter ([\[1\]](https://github.com/pytorch/pytorch/issues/47915) [\[2\]](https://github.com/pytorch/pytorch/issues/43552) [\[3\]](https://github.com/pytorch/pytorch/issues/36214) [\[4\]](https://github.com/pytorch/pytorch/issues/32928) [\[5\]](https://discuss.pytorch.org/t/pruning-doesnt-affect-speed-nor-memory-for-resnet-101/75814) [\[6\]](https://discuss.pytorch.org/t/discussion-of-practical-speedup-for-pruning/75212) [\[7\]](https://discuss.pytorch.org/t/save-pruned-model-after-pruning/103212) [\[8\]](https://stackoverflow.com/questions/62326683/prunning-model-doesnt-improve-inference-speed-or-reduce-model-size) and so on.)

## Pitch

We (me and **[@AndreaBrg](https://github.com/AndreaBrg)**) have currently developed a set of utilities for simplifying pruned models, and they could be ported to the pytorch core library with minor modifications. We opened this issue to gather feedback about your potential interest.

## Additional context

At the time being our codebase works with most of the torchvision models, and we observe noticeable improvements in the inference times (10 runs on cpu, with a batch size of 100, and a structured pruning ratio of 50%): results are reported in the table attached below. *Pruned time* refers to the inference time of the zeroed-out model (as obtained by prune.random_structured and prune.remove), *Simplified time* is the inference time of the corresponding simplified model. Of course, given that we actually remove entire portions of the weight matrices, we also obtain a gain in memory and storage occupation.

If you are interested in our proposal, we can further explain the technical details of our implementation.

| Architecture   | Pruned time    | Simplified time (p=50%)   |
|----------------|----------------|---------------------------|
| alexnet        | 0.2414s ± 0.0065 | 0.1064s ± 0.0010            |
| resnet18       | 1.0082s ± 0.0108 | 0.5528s ± 0.0080            |
| resnet34       | 1.6838s ± 0.0037 | 0.8636s ± 0.0022            |
| resnet50       | 3.6443s ± 0.0029 | 2.1050s ± 0.0034            |
| resnet101      | 5.5948s ± 0.0166 | 3.1392s ± 0.0033            |
| resnet152      | 7.8656s ± 0.0405 | 4.3677s ± 0.0059            |
| vgg16          | 4.8807s ± 0.0210 | 1.9711s ± 0.0272            |
| vgg16_bn       | 6.4215s ± 0.0063 | 1.9731s ± 0.0068            |
| vgg19          | 5.9035s ± 0.0445 | 2.2855s ± 0.0079            |
| vgg19_bn       | 7.5785s ± 0.0372 | 2.2698s ± 0.0106            |

**Comment by albanD** [comment]
Hi,

That sounds very interesting indeed.
Could you give some details how you do that? Or maybe share some code that you use to do this?

**Comment by carloalbertobarbano** [comment]
Thanks for the interest, of course. 

The process consists of three steps:

1. Fusion of BatchNorm layers
2. Propagation of biases
3. Removal of zeroed tensor channels

We are now working on a writeup to explain the maths, which we will share it soon along with the code

**Comment by albanD** [comment]
That sounds good. Thanks!

Just to be sure you have the right expectations and you know the kind of information we're looking for here.
The policy we have is to only include in pytorch core code that is generic for all users.
And that means that logic that is more specific to vision models for example should go to torchvision.
Or logic that is specific to a type of application should stay in a user's third party repo.

And of course if this third party repo starts to be used more broadly, we are happy to re-consider moving it into core/a domain library.

**Comment by carloalbertobarbano** [comment]
Thanks for the clarification. Yes, the code we implemented should actually be generic enough to be usable by all users (and in our opinion and for our use-case it made sense to merge it into the pytorch pruning module).

**Comment by AndreaBrg** [comment]
We have been working on the simplification code, you can find it here: https://github.com/EIDOSlab/simplify.  The core code can be found in the "simplify" module, which takes care of the three steps mentioned above.

Also, here is a writeup for the maths involved in the process https://github.com/EIDOSlab/simplify/blob/master/simplification.pdf, this should make following the code flow easier. 

In the README and in the [profile folder](https://github.com/EIDOSlab/simplify/tree/master/profile) file you can find results with all of the compatible torchvision models (with the exception of densenet which is the only architecture still causing some issues). 

Let us know what you think, and if you consider this contribution suited to be incorporated into pytorch we can start working on an actual PR. Thanks

**Comment by albanD** [comment]
Thanks for the materials, let me sync with the other maintainers and get back to you!

**Comment by z-a-f** [comment]
This is cool stuff. I think we have a similar effort going on right now, but in a little more generic terms (we are also thinking about training time implications).

This is cool stuff, thank you for doing this.

If you want to create a PR -- this is a little tricky. We are adding another package within pytorch (`torch.ao`), which would contain the tools for architecture optimization. However, this is still WIP. Let's keep this conversation going though.

cc @raghuramank100 @kazhou

**Comment by AndreaBrg** [comment]
Hi, thanks for considering our material interesting.

We are interested, if possible, to help with the development of this feature and eventualy draft a PR.
Is there a specific branch we should look for the torch.ao package? Currently on the master branch it looks more focused on quantizations instead of pruning.

Also we've seen that https://dev-discuss.pytorch.org/ exists, would it be better to carry on the conversation there?

**Comment by z-a-f** [comment]
@AndreaBrg Let us put together a simple plan of action, and we can discuss later on how we sync your work with the existing one.

Thank you

**Comment by albanD** [comment]
From taking a closer look, these are the main features that you propose:
- Fold conv+batchnorm into a single conv
- Propagate bias into weights for linear, conv and batchnorm and propagate to next module when possible
- remove zeroed parameters in conv, linear and batchnorm
Did I miss anything @AndreaBrg ?

@z-a-f are these transforms also in the list of things you are planning to do?

A major difference I see here is that all these changes are done directly using torch.nn features (not using fx) while I think you were planning to do all your transforms based on fx @z-a-f ?

**Comment by AndreaBrg** [comment]
Hi @albanD

Yes, those are overall the main steps we propose, plus some other minor details needed to handle skip and residual connections.

We decided to use `torch.nn` because we have no experience with the `torch.fx` module, but from a quick look it should be feasible to convert each step using `torch.fx`.

**Comment by albanD** [comment]
There are definitely tradeoff in using torch.fx as well as it doesn't handle all Modules. So keeping this as a pure nn construct would also be interesting here.

**Comment by carloalbertobarbano** [comment]
Hello, is there any update on this issue?

**Comment by albanD** [comment]
cc @z-a-f ?

**Comment by Laubeee** [comment]
Seems the efforts have not been further pursued..
Is there any chance to have the simplify repository actively maintained somehow?

**Comment by AndreaBrg** [comment]
@Laubeee @albanD  the development and maintenance of simplify was put on hold due to changes in work and life priorities.

That said we'd be happy to pick it back up given enough intereset (we should at least check if newer PyTorch releases are still compatible) and we are open to PRs.
